### PR TITLE
Make it easier to author a hat file from python

### PR DESCRIPTION
**Pull request by masonremy**
- Instantiate more default objects for simple cases
- Enable modifying the list of functions in the hat file
- Make hat file functions and function_map wrap the same underlying
  object

### CODE REVIEW

File: hatlib/hat_file.py [hunk 1]
```diff
@@ -544,33 +560,86 @@ class HATFile:
...
+        return self._device_function_table.function_map
+
+    @device_function_map.setter
+    def device_function_map(self, func_map):
+        if isinstance(func_map, property):
+            return
+        self._device_function_table.function_map = func_map
 
     def Serialize(self, filepath=None):
         """Serilizes the HATFile to disk using the file location specified by `filepath`.
```
Comment (lisaong) [inline hatlib/hat_file.py:645]: [nit] typo

File: hatlib/hat_file.py [hunk 2]
```diff
@@ -544,33 +560,86 @@ class HATFile:
...
             func.hat_file = self
             func.link_target = Path(self.path).resolve().parent / self.dependencies.link_target
 
-        if not self._device_function_table:
-            self._device_function_table = DeviceFunctionTable({})
-        self.device_function_map = self._device_function_table.function_map
-        self.device_functions = self._device_function_table.functions
+    @property
+    def functions(self):
+        if self._function_table is None:
```
Comment (lisaong) [inline hatlib/hat_file.py:586]: not sure if we'd ever hit the None case, since there is a default initializer and the setter doesn't handle None.

**Review by lisaong — APPROVED** [review]

File: test/test_create_simple_hat_file.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+        self.assertTrue(func_name in parsed_hat_file.function_map)
+        self.assertEqual(parsed_hat_file.dependencies.link_target, link_target_path)
+        self.assertEqual(len(parsed_hat_file.function_map[func_name].arguments), 3)
+        self.assertEqual(parsed_hat_file.function_map[func_name].arguments[0].name, param_A.name)
+        self.assertEqual(parsed_hat_file.function_map[func_name].arguments[1].name, param_B.name)
+        self.assertEqual(parsed_hat_file.function_map[func_name].arguments[2].name, param_C.name)
+        self.assertEqual(parsed_hat_file.function_map[func_name].arguments[0].shape, list(param_A.shape))
+        self.assertEqual(parsed_hat_file.function_map[func_name].arguments[1].shape, list(param_B.shape))
+        self.assertEqual(parsed_hat_file.function_map[func_name].arguments[2].shape, list(param_C.shape))
```
Comment (kernhanda) [inline test/test_create_simple_hat_file.py:126]: do you want to also assert that the deserialized decl matches `SAMPLE_MATMUL_DECL_CODE`?

**Review by kernhanda — APPROVED** [review]
lgtm

**Pull request merged**